ivy.el

```diff
@@ -3119,11 +3138,221 @@ This function serves as a fallback when nothing else is available."
+
+(defun ivy--fuzzy-plan (pat pat-role pat-set text text-role text-set miss match best-pos)
+  "Return a list of positions in TEXT after a fuzzy match with PAT.
+
+PAT-ROLE PAT-SET TEXT-ROLE TEXT-SET are used for scoring heuristics.
+MISS MATCH are 2D tables.
+BEST-POS is the last matching position in TEXT of the best plan."
+  (cl-macrolet ((c0 (i j) `(aref (aref miss ,i) ,j))
+                (c1 (i j) `(aref (aref match ,i) ,j)))
+    (let* (plan
+           (i (length pat))
+           (j best-pos)
+           (is-match t))
+      (while (> i 0)
+        (if is-match
+            ;; pat[i] matches text[i]. Trace back to pat[i-1] and text[j-1].
+            (progn
+              (setq is-match
+                    (= (+ (c1 (1- i) (1- j))
+                          (ivy--fuzzy-match-score
+                           (1- i) (1- j) t pat pat-role pat-set text text-role))
+                       (c1 i j)))
+              (cl-decf i)
+              (push (1- j) plan))
+          ;; Trace back to pat[i] and text[j-1].
+          (setq is-match
+                (= (+ (c1 i (1- j))
+                      (ivy--fuzzy-miss-score (1- j) t text-role))
+                   (c0 i j))))
+        (cl-decf j))
+      (nreverse plan))))
+
+(defun ivy--fuzzy-sort (pat cands)
+  "Sort according to closeness to string PAT the string list CANDS."
+  (let ((regex (mapconcat
+                (lambda (x)
+                  (format "\\(%s\\)" (regexp-quote (string x))))
+                (delete ?\  (string-to-list pat))
+                ".*?")))
+    (setq cands (cl-delete-if-not (apply-partially #'string-match regex) cands)))
+  (let* ((m (length pat))
+         (n (apply #'max 0 (mapcar #'length cands)))
+         pat-set
+         text-set
+         ;; Vectors of roles for each position of pat and text.
+         (pat-role (make-vector m nil))
+         (text-role (make-vector n nil))
+         ;; match and miss are 2D DP vectors.
+         (miss (make-vector (1+ m) nil))
+         (match (make-vector (1+ m) nil))
+         cands-to-sort
+         cands-with-score
+         plan)
+    (setq pat-set (ivy--fuzzy-get-role pat pat-role))
+    (setq m 0)
+    (dotimes (i (length pat))
+      (when (/= (aref pat i) ?\ )
+        (aset pat m (aref pat i))
+        (aset pat-role m (aref pat-role i))
+        (cl-incf m)))
+    (setq pat (substring-no-properties pat 0 (min m ivy-fuzzy-pattern-length-limit)))
+
+    (dotimes (i (1+ m))
+      (setf (aref miss i) (make-vector (1+ n) 0)
+            (aref match i) (make-vector (1+ n) 0)))
+
+    ;; Partition the candidates into sorted and unsorted groups.
+    (dotimes (_i (min n ivy-fuzzy-limit))
+      (push (pop cands) cands-to-sort))
+
+    ;; For each candidate, compute the score.
+    (dolist (text cands-to-sort)
+      (if (> (length text) ivy-fuzzy-candidate-length-limit)
+          (setq text (substring text 0 ivy-fuzzy-candidate-length-limit))
+        (setq text (copy-sequence text)))
```

Comment:
Unless you want to avoid allocating a new string in the alternative branch, why not just write
```el
(setq text (substring text 0 (min (length text)
                                  ivy-fuzzy-candidate-length-limit)))
```